src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/CodeFoldingTests.cs

```diff
@@ -0,0 +1,186 @@
+
+            var outlines = manager.GetAllRegions(span);
+
+            return outlines
+                    .OrderBy(s => s.Extent.GetStartPoint(textView.TextSnapshot))
+                    .ToArray();
+        }
+
+        private async Task AssertFoldableBlocksAsync(Span[] foldableSpans)
+        {
+            var textView = await TestServices.Editor.GetActiveTextViewAsync(HangMitigatingCancellationToken);
+            var outlines = await GetOutlineRegionsAsync(textView);
+
+            //
+            // Built in try logic because getting spans can take time.
+            //
+            var tries = 0;
+            const int MaxTries = 10;
+            ImmutableArray<Span> missingSpans;
+            while (tries++ < MaxTries)
+            {
+                missingSpans = GetMissingExpectedSpans(outlines, foldableSpans, textView);
+                if (missingSpans.Length == 0)
+                {
+                    return;
+                }
+
+                Thread.Sleep(100);
+            }
+
+            Assert.Empty(missingSpans);
+
+            static ImmutableArray<Span> GetMissingExpectedSpans(ICollapsible[] outlines, Span[] foldableSpans, ITextView textView)
+            {
+                if (outlines.Length < foldableSpans.Length)
+                {
+                    return foldableSpans.ToImmutableArray();
+                }
+
+                var builder = new List<Span>();
+
+                var spans = outlines.Select(o => o.Extent.GetSpan(textView.TextSnapshot).Span).ToImmutableArray();
+                foreach (var foldableSpan in foldableSpans)
+                {
+                    if (!spans.Contains(foldableSpan))
+                    {
+                        builder.Add(foldableSpan);
+                    }
+                }
+
+                return builder.ToImmutableArray();
+            }
+        }
+
+        [IdeFact]
+        public async Task CodeFolding_CodeBlock_Default()
+        {
+            await TestServices.SolutionExplorer.OpenFileAsync(BlazorProjectName, CounterRazorFile, HangMitigatingCancellationToken);
+
+            await AssertFoldableBlocksAsync(new Span[]
+            {
+                Span.FromBounds(285, 324)
+            });
+        }
+
+        [IdeFact]
+        public async Task CodeFolding_IfBlock()
+        {
+            await TestServices.SolutionExplorer.AddFileAsync(
+                BlazorProjectName,
+                "Test.razor",
+                @"
+@page ""/Test""
+
+<PageTitle>Test</PageTitle>
+
+<h1>Test</h1>
+
+@if(true)
+{
+    if (true)
+    {
+        M();
+    }
+}
+
+@code {
+    string M() => ""M"";
+}
+",
+                open: true,
+                HangMitigatingCancellationToken);
+
+            await AssertFoldableBlocksAsync(new Span[]
+            {
+                Span.FromBounds(76, 125), // Outer @if
+                Span.FromBounds(94, 122)  // Inner if
```

Comment:
Oooooo one thing I just realized, because we don't rely on TextMate anymore we're now losing folding ranges for `@code` and `@{...}`. Aka:
![image](https://user-images.githubusercontent.com/2008729/156109129-c6640b72-ade5-4a6d-8c3a-d7c51fe560eb.png)

Crap, we may need to build a system for that to include those ranges in 17.2. Hmmm, this puts us in an interesting position on when we should land these changes, let sync-up tomorrow (3/1)